### PR TITLE
Next Word Prediction using LSTM

### PR DESCRIPTION
**Pull request by snehapm04**
Closes: #97 

- **Title** : Next Word Prediction using LSTM
- **Your Name:** Sneha P M
- **Identify yourself:** SWOC Contributor

### Describe the add-ons or changes you've made 📃

> I have added the following enhancements to the project

1. Imported and structured the necessary libraries for text preprocessing, sequence creation, and model development using TensorFlow and Keras.
2. Loaded and preprocessed the dataset by cleaning text, tokenizing sentences, and generating input-output word sequences.
3. Visualized word distributions and sequence lengths to understand the dataset's structure and potential challenges.
4. Designed and trained an LSTM-based neural network for next-word prediction, using embedding layers and sequential modeling techniques.
5. Evaluated model performance using metrics like categorical cross-entropy loss and accuracy, and visualized training and validation loss trends.

### Checklist: ☑️

- [x] My code follows the [Contributing Guidelines](https://github.com/Avdhesh-Varshney/AI-Code/blob/main/README.md) & [Code of Conduct](https://github.com/Avdhesh-Varshney/AI-Code/blob/main/CODE_OF_CONDUCT.md) of this project.
- [x] This PR does not contain plagiarized content.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly wherever it was hard to understand.
- [x] My changes generate no new warnings.

### Screenshots 📷

image
image
model perf
Screenshot 2025-01-07 121548

### Working Video 🎥

[Next Word Prediction Recording](https://drive.google.com/file/d/1jWKM0E5hD4MvRHkJtppHXVbznTp-lSS5/view?usp=sharing)

### CODE REVIEW

**Review by Avdhesh-Varshney — APPROVED** [review]
Really Appreaciable Work!

File: docs/NLP/projects/next-word-pred.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+### AIM 
+To predict the next word using LSTM.
+
+
+### DATASET LINK 
+[https://www.kaggle.com/datasets/muhammadbilalhaneef/sherlock-holmes-next-word-prediction-corpus](https://www.kaggle.com/datasets/muhammadbilalhaneef/sherlock-holmes-next-word-prediction-corpus)
+
+### NOTEBOOK LINK 
+[https://colab.research.google.com/drive/1w7PvMVj7U_sTdVnJHWrgTK68Gb6XPsqG?usp=sharing](https://drive.google.com/file/d/1G4XD7VCkI4z6OnZ2vC7qQgxoY3hD2dyY/view?usp=sharing)
```
[the remark this answers is not in the document]
Comment (11PRIMUS) [reply docs/NLP/projects/next-word-pred.md:13]: remove the link of first  one and in second one use bidirectional lstm or you can use different datasets to improve accuracy

[the remark this answers is not in the document]
Comment (snehapm04) [reply docs/NLP/projects/next-word-pred.md:13]: Ok

[the remark this answers is not in the document]
Comment (11PRIMUS) [reply docs/NLP/projects/next-word-pred.md]: @snehapm04  use earlystopping to halt training when improvements plateau, and reduce the epoch size it can crash the memory

[the remark this answers is not in the document]
Comment (snehapm04) [reply docs/NLP/projects/next-word-pred.md]: @11PRIMUS I have made these changes in the code.

**Review by 11PRIMUS — APPROVED** [review]

**Review by Avdhesh-Varshney — APPROVED** [review]
